### PR TITLE
Added ICEADMM, IIADMM, and client weight computation

### PR DESCRIPTION
**Pull request by minseok-ryu**
Add ICEADMM

### CODE REVIEW

File: appfl/run.py [hunk 1]
```diff
@@ -13,25 +14,24 @@
 import time
 from .misc.data import Dataset
 from .misc.utils import *
-from .algorithm.iadmm import *
+ 
 from .algorithm.fedavg import *
-
+from .algorithm.iceadmm import *
+from .algorithm.iiadmm import *
```
Comment (kibaekkim) [inline appfl/run.py:21]: Instead of adding every single module here, can we do something like `from .algorithm import *`?

Comment (minseok-ryu) [reply appfl/run.py:21]: I have tried it, but it is not working.

**Comment by minseok-ryu** [comment]
Ready.

File: appfl/algorithm/fedavg.py [hunk 2]
```diff
@@ -4,71 +4,97 @@
...
+
+        self.primal_states = OrderedDict()
+        for i in range(num_clients):
+            self.primal_states[i] = OrderedDict()
+
+    def update(self, local_states: OrderedDict):
+
+        """ Inputs """
+        global_state = OrderedDict()
+        primal_recover_from_local_states(self, local_states)
```
Comment (kibaekkim) [inline appfl/algorithm/fedavg.py:32]: It seems this function is used in derived algorithms. It should have been defined in `BaseServer`.

File: appfl/algorithm/fedavg.py [hunk 3]
```diff
@@ -4,71 +4,97 @@
...
+    def __init__(self, weights, model, num_clients, device, **kwargs):
+        super(FedAvgServer, self).__init__(weights, model, num_clients, device)
 
         self.__dict__.update(kwargs)
+        self.num_clients = num_clients
+        self.weights = weights
+
+        self.primal_states = OrderedDict()
+        for i in range(num_clients):
+            self.primal_states[i] = OrderedDict()
```
Comment (kibaekkim) [inline appfl/algorithm/fedavg.py:26]: These variables seem used in other algorithm derivations. Any reason that they are defined locally? not in `BaseServer`?

Comment (kibaekkim) [inline appfl/algorithm/fedavg.py:26]: This variable seems used in other algorithm derivations. Any reason that this is defined locally? not in `BaseServer`?

File: appfl/algorithm/fedavg.py [hunk 4]
```diff
@@ -4,71 +4,97 @@
...
-    # update global model
-    def update(self, global_state: OrderedDict, local_states: OrderedDict):
-        update_state = OrderedDict()
-        for k, state in local_states.items():
-            for key in self.model.state_dict().keys():
-                if key in update_state.keys():
-                    update_state[key] += state[key] / self.num_clients
-                else:
-                    update_state[key] = state[key] / self.num_clients
+        """ Outputs """
```
Comment (kibaekkim) [inline appfl/algorithm/fedavg.py:35]: Inputs and Outputs..? Can you give us some more descriptions?

File: appfl/algorithm/fedavg.py [hunk 5]
```diff
@@ -4,71 +4,97 @@
...
+        if self.privacy == True:
+            # Note: Scale_value = Sensitivity_value / self.epsilon
+
+            Scale_value = self.scale_value
+
+            for name, param in self.model.named_parameters():
+                mean = torch.zeros_like(param.data)
+                scale = torch.zeros_like(param.data) + Scale_value
+                m = torch.distributions.laplace.Laplace(mean, scale)
+                param.data += m.sample()
```
Comment (kibaekkim) [inline appfl/algorithm/fedavg.py:92]: This looks like output perturbation, correct? Please leave some descriptions about DP.

File: appfl/algorithm/iceadmm.py [hunk 6]
```diff
@@ -0,0 +1,156 @@
...
+        self.num_clients = num_clients
+        self.weights = weights
+
+        self.primal_states = OrderedDict()
+        self.dual_states = OrderedDict()
+        for i in range(num_clients):
+            self.primal_states[i] = OrderedDict()
+            self.dual_states[i] = OrderedDict()
+
+        self.penalty = OrderedDict()
```
Comment (kibaekkim) [inline appfl/algorithm/iceadmm.py:30]: Please consider moving these under `BaseServer`.

File: appfl/algorithm/iceadmm.py [hunk 7]
```diff
@@ -0,0 +1,156 @@
...
+
+    def update(self, local_states: OrderedDict):
+        
+        """ Inputs """
+        global_state = self.model.state_dict()
+        primal_recover_from_local_states(self, local_states)
+        dual_recover_from_local_states(self, local_states)
+        penalty_recover_from_local_states(self, local_states)
```
Comment (kibaekkim) [inline appfl/algorithm/iceadmm.py:38]: Please consider moving these under `BaseServer`. And, these functions do pretty much the same thing.

File: appfl/algorithm/misc.py [hunk 8]
```diff
@@ -0,0 +1,83 @@
+import copy
+
+"""Functions for server."""
```
Comment (kibaekkim) [inline appfl/algorithm/misc.py:3]: I think these functions can be defined under proper classes....

File: appfl/run.py [hunk 9]
```diff
@@ -136,13 +140,25 @@ def run_server(
...
+    for rank in range(1, comm_size):
+        for val in Num_Data[rank].values():
+            total_num_data += val
+    weights = {}
+    for rank in range(1, comm_size):
+        weight = {}
+        for key in Num_Data[rank].keys():
+            weight[key] = Num_Data[rank][key] / total_num_data
+            weights[key] = weight[key]
+        comm.send(weight, dest=rank)
```
Comment (kibaekkim) [inline appfl/run.py:154]: use scatter

**Review by kibaekkim — COMMENTED** [review]
Do we also need to update gRPC?? @youngdae

File: appfl/algorithm/fedavg.py (hunk 2 above)
Comment (minseok-ryu) [reply appfl/algorithm/fedavg.py:32]: Fixed

File: appfl/algorithm/fedavg.py (hunk 3 above)
Comment (minseok-ryu) [reply appfl/algorithm/fedavg.py:26]: Fixed

File: appfl/algorithm/fedavg.py (hunk 4 above)
Comment (minseok-ryu) [reply appfl/algorithm/fedavg.py:35]: Fixed

File: appfl/algorithm/fedavg.py (hunk 5 above)
Comment (minseok-ryu) [reply appfl/algorithm/fedavg.py:92]: Fixed. I added the description on "BaseClient"

File: appfl/algorithm/iceadmm.py (hunk 6 above)
Comment (minseok-ryu) [reply appfl/algorithm/iceadmm.py:30]: Reflected

File: appfl/algorithm/iceadmm.py (hunk 7 above)
Comment (minseok-ryu) [reply appfl/algorithm/iceadmm.py:38]: Reflected

File: appfl/algorithm/misc.py (hunk 8 above)
Comment (minseok-ryu) [reply appfl/algorithm/misc.py:3]: misc.py is removed

File: appfl/run.py (hunk 9 above)
Comment (minseok-ryu) [reply appfl/run.py]: reflected

**Comment by youngdae** [comment]
There's no update on gRPC, yet. I'm working on improving the performance of gRPC. From the literature, it seems that streaming gRPC on Python is expected to be slow (https://grpc.io/docs/guides/performance/). I'm checking ways to get around this issue.

**Comment by youngdae** [comment]
> Do we also need to update gRPC?? @youngdae
There's no update on gRPC, yet. I'm working on improving the performance of gRPC. From the literature, it seems that streaming gRPC on Python is expected to be slow (https://grpc.io/docs/guides/performance/). I'm checking ways to get around this issue.

File: appfl/algorithm/iiadmm.py [hunk 10]
```diff
@@ -141,13 +141,14 @@ def update(self):
             super(IIADMMClient, self).laplace_mechanism_output_perturb()
 
         """ Update local_state """
-        self.local_state = OrderedDict()                
-        self.local_state["primal"] = copy.deepcopy(self.primal_state)                        
+        self.local_state = OrderedDict()
+        self.local_state["primal"] = copy.deepcopy(self.primal_state)
+        self.local_state["dual"] = OrderedDict()
```
Comment (kibaekkim) [inline appfl/algorithm/iiadmm.py:146]: @minseok-ryu please confirm that this change is ok.

**Review by kibaekkim — APPROVED** [review]

File: appfl/algorithm/iiadmm.py (hunk 10 above)
Comment (minseok-ryu) [reply appfl/algorithm/iiadmm.py:146]: It is okay!

**Pull request merged**